src/pipeline-stages/src/main/scala/TextPreprocessor.scala

```diff
@@ -0,0 +1,134 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import breeze.linalg.max
+import org.apache.spark.ml.Transformer
+import org.apache.spark.ml.param.{MapParam, ParamMap}
+import org.apache.spark.ml.util.{ComplexParamsReadable, ComplexParamsWritable, Identifiable}
+import org.apache.spark.sql.types.{StringType, StructField, StructType}
+import org.apache.spark.sql.{DataFrame, Dataset}
+import org.apache.spark.sql.functions.udf
+import spray.json.DefaultJsonProtocol._
+
+import scala.collection.mutable
+
+class Trie(normFunc: String => String = x => x) extends Serializable {
+  var trieMap = mutable.Map[String, Trie]()
+  var normFunction = normFunc
+  var value = ""
+
+  def put(key:String, value:String): Trie =  (key, value) match {
+    case (_, _) if key.length() == 0 => this
+    case (k, v) => {
+      val normK: String = normFunction(k)
+      val first = normK.substring(0, 1)
+      val next: Trie = trieMap.getOrElse(first, new Trie()).put(normK.substring(1), v)
+      if (normK.length() == 1) next.value = v
+      this.trieMap(normK.substring(0, 1)) = next
+      this
+    }
+  }
+
+  def get(letter:String): Trie = trieMap.getOrElse(letter, null)
```

Comment:
don't use `null` use `Option[String]` intstead

Reply:
Made the change. The get function for Options was a bit odd, but the solution is more robust now. Thanks